lib/modules/platform/scm-manager/utils.ts

```diff
@@ -0,0 +1,101 @@
+import type { MergeStrategy } from '../../../config/types';
+import { logger } from '../../../logger';
+import { regEx } from '../../../util/regex';
+import { parseUrl } from '../../../util/url';
+import type { GitUrlOption, Pr } from '../types';
+import type { Link, PRMergeMethod, PrFilterByState, Repo } from './types';
+
+export function mapPrState(
+  state: 'open' | 'closed' | undefined,
+): 'OPEN' | 'REJECTED' | undefined {
+  switch (state) {
+    case 'open':
+      return 'OPEN';
+    case 'closed':
+      return 'REJECTED';
+    default:
+      return undefined;
+  }
+}
+
+export function matchPrState(pr: Pr, state: PrFilterByState): boolean {
+  if (state === 'all') {
+    return true;
+  }
+
+  if (state === 'open' && (pr.state === 'OPEN' || pr.state === 'DRAFT')) {
+    return true;
+  }
+
+  if (state === '!open' && (pr.state === 'MERGED' || pr.state === 'REJECTED')) {
+    return true;
+  }
+
+  if (
+    state === 'closed' &&
+    (pr.state === 'MERGED' || pr.state === 'REJECTED')
+  ) {
+    return true;
+  }
+
+  return false;
+}
+
+export function smartLinks(body: string): string {
+  return body.replace(regEx(/\]\(\.\.\/pull\//g), '](pulls/');
+}
+
+export function getRepoUrl(
+  repo: Repo,
+  gitUrl: GitUrlOption | undefined,
+  username: string,
+  password: string,
+): string {
+  const protocolLinks = repo._links.protocol as Link[] | undefined;
```

Comment:
```suggestion
  const protocolLinks = repo._links.protocol;
```

should not nesessary to cast here. fix the type